akka-http2-support/src/main/scala/akka/http/impl/engine/http2/WrappedSslContextSPI.scala

```diff
@@ -6,49 +6,87 @@ package akka.http.impl.engine.http2
+      def select(protocols: ju.List[String]): String =
+        choose(protocols.asScala.toSet)
+
+      def unsupported(): Unit =
+        choose(Set.empty)
+
+      def choose(protocols: immutable.Set[String]): String =
+        try decider(protocols) match {
+          case entry @ Some(chosen) ⇒
+            if (protocols.nonEmpty && !protocols(chosen))
+              throw new SSLException(s"Tried to choose unsupported protocol. Chosen $chosen supported: [$protocols]")
+            else {
+              println(s"Chose: $chosen for engine ${engine}")
+              putSelection(engine, entry)
+              chosen
+            }
+          case None ⇒
+            println(s"Chose nothing")
+            putSelection(engine, None)
+
+            if (protocols.nonEmpty)
+              throw new SSLException(s"No protocol chosen from [$protocols]")
+            else ""
+        }
+        finally ALPN.remove(engine)
+    })
+
+  def getChosenProtocolAndRemove(engine: SSLEngine): Option[String] = synchronized {
+    val selected = selections.getOrElse(engine, None)
+    selections -= engine
+    selected
+  }
+
+  var selections = Map.empty[SSLEngine, Option[String]]
+  private def putSelection(engine: SSLEngine, selection: Option[String]): Unit =
+    synchronized {
+      selections += (engine → selection)
+    }
+}
 
 /**
  * Wraps an SSLContext so that all engines it creates will announce HTTP/2 if possible.
  */
 class WrappedSslContextSPI(underlying: SSLContext) extends SSLContextSpi {
+  // this context is only allowed to create a single engine
+  @volatile private[this] var theEngine: SSLEngine = null
+  @volatile private[this] var theStacktrace: Throwable = null
+
+  def getSingleEngine: SSLEngine =
+    if (theEngine ne null) theEngine
+    else throw new IllegalStateException("SSLEngine was not set yet")
+
   def engineGetSocketFactory(): SSLSocketFactory =
     underlying.getSocketFactory()
 
   def engineInit(keyManagers: Array[KeyManager], trustManagers: Array[TrustManager], secureRandom: SecureRandom): Unit =
     underlying.init(keyManagers, trustManagers, secureRandom)
 
-  def engineCreateSSLEngine(): SSLEngine = {
-    val engine = underlying.createSSLEngine()
-    ALPN.put(engine, new ALPN.ServerProvider {
-      def select(protocols: ju.List[String]): String =
-        try
-          if (protocols.contains("h2")) {
-            System.out.println("HTTP/2 is supported!")
-            "h2"
-          } else
-            // FIXME: actually negotiate and fall back to HTTP/1.1
-            throw new SSLException("No supported protocol found. Only 'h2' supported right now.")
-        finally ALPN.remove(engine)
-
-      def unsupported(): Unit = {
-        println("ALPN not supported by client!") // FIXME: remove
-        ALPN.remove(engine)
+  def engineCreateSSLEngine(): SSLEngine =
+    if (theEngine == null) {
+      val engine = underlying.createSSLEngine()
+      ALPNRegistry.registerServerEngine(engine) { protocols ⇒
```

Comment:
Yes, we need some kind of side-channel as the ALPN handler thing. But right now I figured that this is a much more complicated side-channel than we actually need. I'll update with something much simpler based on a volatile var.